app/models/puppetclass.rb

```diff
@@ -87,31 +84,25 @@ def klass
     name.gsub(module_name+"::","")
   end
 
-  # return host ids from config groups by type
-  def host_ids_from_config_groups(host_type)
-    ids = config_groups.joins(:host_config_groups)
-                 .where("host_config_groups.host_type='#{host_type}'")
-                 .pluck('host_config_groups.host_id') unless config_group_classes.empty?
-    ids || []
-  end
-
   def all_hostgroups(with_descendants = true)
-    ids = hostgroup_ids
-    ids += host_ids_from_config_groups('Hostgroup')
-    hgs = Hostgroup.unscoped.where(:id => ids.uniq)
+    hgs = Hostgroup.authorized
+                   .reorder('')
+                   .eager_load(:hostgroup_classes, :config_groups => [:config_group_classes])
+                   .where("#{self.id} IN (hostgroup_classes.puppetclass_id, config_group_classes.puppetclass_id)")
+                   .uniq
     hgs = hgs.flat_map(&:subtree).uniq if with_descendants
     hgs
   end
 
-  def all_host_ids
-    ids = host_ids
-    ids += Host::Managed.unscoped.where(:hostgroup_id => all_hostgroups.map(&:id)).pluck(:id)
-    ids += host_ids_from_config_groups('Host::Base')
-    ids.uniq
-  end
-
-  def update_total_hosts
-    update_attribute(:total_hosts, all_host_ids.size)
+  def hosts_count
+    ids = Host::Managed.authorized
+                       .reorder('')
+                       .eager_load(:host_classes, :config_groups => [:config_group_classes])
+                       .where("#{self.id} IN (host_classes.puppetclass_id, config_group_classes.puppetclass_id)")
+                       .uniq
+                       .pluck(:id)
+    ids |= Host::Managed.authorized.where(:hostgroup_id => all_hostgroups.map(&:id)).pluck(:id)
+    ids.size
   end
```

Comment:
`|=` is set union for arrays - I think you confused it with `||=`. This is because we might get duplicate hosts in the two queries - which is why we can't count in them either. Anyways, I'll change these two queries to run in one, with a bit of extra sql.